67_RayQueryGeometry/main.cpp

```diff
@@ -0,0 +1,998 @@
+				blasBuildInfo.buildFlags = blasFlags;
+				blasBuildInfo.geometryCount = 1;	// only 1 geometry object per blas
+				blasBuildInfo.srcAS = nullptr;
+				blasBuildInfo.dstAS = nullptr;
+				blasBuildInfo.triangles = &triangles;
+				blasBuildInfo.scratch = {};
+
+				ILogicalDevice::AccelerationStructureBuildSizes buildSizes;
+				{
+					const uint32_t maxPrimCount[1] = { trisCount };
+					buildSizes = m_device->getAccelerationStructureBuildSizes(blasFlags, false, std::span{ &triangles, 1 }, maxPrimCount);
+				}
+
+				{
+					IGPUBuffer::SCreationParams params;
+					params.usage = bitflag(IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT) | IGPUBuffer::EUF_ACCELERATION_STRUCTURE_STORAGE_BIT;
+					params.size = buildSizes.accelerationStructureSize;
+					smart_refctd_ptr<IGPUBuffer> asBuffer = createBuffer(params);
+
+					IGPUBottomLevelAccelerationStructure::SCreationParams blasParams;
+					blasParams.bufferRange.buffer = asBuffer;
+					blasParams.bufferRange.offset = 0u;
+					blasParams.bufferRange.size = buildSizes.accelerationStructureSize;
+					blasParams.flags = IGPUBottomLevelAccelerationStructure::SCreationParams::FLAGS::NONE;
+					gpuBlas = m_device->createBottomLevelAccelerationStructure(std::move(blasParams));
+				}
+
+				smart_refctd_ptr<IGPUBuffer> scratchBuffer;
+				{
+					IGPUBuffer::SCreationParams params;
+					params.usage = bitflag(IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT) | IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+					params.size = buildSizes.buildScratchSize;
+					scratchBuffer = createBuffer(params);
+				}
+
+				blasBuildInfo.dstAS = gpuBlas.get();
+				blasBuildInfo.scratch.buffer = scratchBuffer;
+				blasBuildInfo.scratch.offset = 0u;
+
+				IGPUBottomLevelAccelerationStructure::BuildRangeInfo buildRangeInfos[1u];
+				buildRangeInfos[0].primitiveCount = trisCount;
+				buildRangeInfos[0].primitiveByteOffset = 0u;
+				buildRangeInfos[0].firstVertex = 0u;
+				buildRangeInfos[0].transformByteOffset = 0u;
+				IGPUBottomLevelAccelerationStructure::BuildRangeInfo* pRangeInfos[1u];
+				pRangeInfos[0] = &buildRangeInfos[0];
+
+				cmdbufBlas->buildAccelerationStructures({ &blasBuildInfo, 1 }, pRangeInfos);
+			}
+
+			{
+				SMemoryBarrier memBarrier;
+				memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::ACCELERATION_STRUCTURE_BUILD_BIT;
+				memBarrier.srcAccessMask = ACCESS_FLAGS::ACCELERATION_STRUCTURE_WRITE_BIT;
+				memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::ACCELERATION_STRUCTURE_BUILD_BIT;
+				memBarrier.dstAccessMask = ACCESS_FLAGS::ACCELERATION_STRUCTURE_READ_BIT;
+				cmdbufBlas->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, { .memBarriers = {&memBarrier, 1} });
+			}
+
+			cmdbufBlas->resetQueryPool(queryPool.get(), 0, 1);
+
+			uint32_t queryCount = 0;
+			const IGPUAccelerationStructure* ases[1u] = { gpuBlas.get() };
+			cmdbufBlas->writeAccelerationStructureProperties({ ases, 1}, IQueryPool::ACCELERATION_STRUCTURE_COMPACTED_SIZE, queryPool.get(), queryCount++);
```

Comment:
if the "queries need no barriers" thing is true, then you don't need a pipeline barrier.